File: leetcode3/이진희/648. Replace Words.java
/* 1차 풀이

1. 아이디어 : 문자열을 공백기준으로 파싱 후, Set에 미리 저장해놓은 dictionary에서 가장 짧은 문자열 있는지 확인후, 있다면 문자열 교체

2. 시간복잡도 : O(S*L^2)

3. 자료구조/알고리즘 : 문자열 계산 , Set

 */

class Solution {
    public String replaceWords(List<String> dictionary, String sentence) {
        // 딕셔너리 포함 단어로 교체, 여러개면 더 짧은걸로
        // 최대 문장길이 100만
        
        String[] words = sentence.split(" ");
        Set<String> set = new HashSet<>();

        for(int i=0; i<dictionary.size(); i++) {
            set.add(dictionary.get(i));
        }

        StringBuilder ans = new StringBuilder();

        // 최대 10만 길이 dictionary
        for(int word=0; word<words.length; word++) {
            StringBuilder sb = new StringBuilder();
            String s = words[word];
            for(int i=0; i<s.length(); i++) {
                sb.append(s.charAt(i));
                if(set.contains(sb.toString())){
                    words[word] = sb.toString();
                    break;
                }
            }
            ans.append(words[word]).append(' ');
        }

        return ans.toString().substring(0, ans.length()-1);
    }
}

File: leetcode3/최원준/648. Replace Words.py
from collections import defaultdict

class Node:
    def __init__(self):
        self.children = {}
        self.word = None

#

'''
1. 아이디어 :
각 단어를 매칭할때 시간복잡도가 터질 수 있다(sentance 단어 1000개, 단어의 길이 1000, dictionary의 단어 1000)
trie 자료 구조를 만들어서 매칭 시간을 줄인다.

2. 시간복잡도 :
    O(a * b + c * d) dictionary.length:a, dictionary.word.length: b, sentence.words:c, word.length:d

3. 자료구조/알고리즘 :
trie

'''
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        trie = Node()

        def insert_word(word, index, node):
            if index >= len(word):
                node.word = word
                return
            char = word[index]
            if char not in node.children:
                child_node = Node()
                node.children[char] = child_node
            else:
                child_node = node.children[char]
            insert_word(word, index+1, child_node)
        
        def search_word(word, index, node):
            if index >= len(word):
                return word
            char = word[index]
            if node.word != None:
                return node.word
            if char not in node.children:
                return word
            child_node = node.children[char]
            return search_word(word, index+1, child_node)        
        
        for word in dictionary:
            insert_word(word, 0, trie)

        words = sentence.split(" ")
        for i in range(len(words)):
            words[i] = search_word(words[i], 0, trie)
        
        return " ".join(words)
        
        
